fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -3,204 +3,683 @@
+
+    def get_batch(self, chunk: int) -> Batch:
+        batch = self.batches[chunk]
+        assert batch is not None
+        return batch
+
+
+class PartitionHandler:
+    """This class processes a single partition of the pipeline.
+    Args:
+        module_rref: RRef to the nn.Module for this partition. It should be on the local rpc worker.
+        device: The device that hols the module.
+        num_input: Numer of inputs to the module
+        num_output: Number of outputs of the module. If the module output is not a tuple (and it is a
+            single tensor), num_output should be None.
+        rank: The rank of the partition
+        chunks: Number of micor-batches in a mini-batch
+        checkpoint_stop::
+    """
+
+    def __init__(
+        self,
+        module_rref: rpc.RRef,
+        device: str,
+        num_input: int,
+        num_output: Optional[int],
+        rank: int,
+        chunks: int,
+        checkpoint_stop: int,
+    ) -> None:
+        self.module = module_rref.local_value()
+        self.chunks = chunks
+        self.device = torch.device(device)
+        self.checkpoint_stop = checkpoint_stop
+        self.rank = rank
+        self.num_input = num_input
+        self.num_output = num_output
+        (self.in_queue,), (self.out_queue,) = create_workers([self.device])
+
+    def local_parameter_rrefs(self) -> List[rpc.RRef]:
+        r"""
+        Create one RRef for each parameter in the given local module, and return a
+        list of RRefs.
+        """
+        return [rpc.RRef(p) for p in self.module.parameters()]
+
+    def make_pipeline_record(self, users: List[Tuple[rpc.RRef, int, int]]) -> DistributedPipelineRecord:
+        return DistributedPipelineRecord(self.device, self.rank, self.chunks, self.num_input, users)
+
+    def run(self, pipeline_record: DistributedPipelineRecord) -> None:
+        """Runs pipeline parallelism.
+
+        It modifies the given batches in place.
+
+        """
+
+        m = len(pipeline_record.batches)
+
+        self.stream = current_stream(self.device)
+
+        for i in range(m):
+            with record_function("feed"):
+                pipeline_record.wait_for(i)
+            self.fence(pipeline_record, i)
+            self.compute(pipeline_record, i)
+            self.forward_results(i, pipeline_record)
+
+    def fence(self, pipeline_record: DistributedPipelineRecord, chunk: int) -> None:
+        """Prepares micro-batches for computation."""
+        # Ensure that batches[chunk-1] is executed after batches[chunk] in
+        # backpropagation by an explicit dependency.
+        # TODO: This dependency injection causes deadlock if this partition
+        # gets its input from model input. 1) Figure out why 2) If we need to live
+        # with this constraint, replace the condition 'pipeline_record.rank > 0' below with
+        # a more accurate one.
+        if chunk != 0 and pipeline_record.users and pipeline_record.rank > 0:
+            t = []
```

Comment:
nit: can we rename variables to be a bit more descriptive? It would be easier to understand the logic.